### PR TITLE
refactor: Fuse initial and final sim particles in Examples

### PR DESCRIPTION
**Pull request by andiwand**
The split of initial and final state of particles became more and more a burden. Currently we have a single EDM to store the particle state after generation and after simulation. While for simulation we have more properties on the particle which is split into `initial` and `final` states. Here I propose to combine the information into the particle object by carrying all states.

Replaces https://github.com/acts-project/acts/pull/3795

## Summary by CodeRabbit

- **New Features**
  - Consolidated output handling for particles by replacing `outputParticlesInitial` and `outputParticlesFinal` with a single `outputParticlesSimulation` parameter across multiple components.
  - Enhanced `SimParticle` class with new methods and properties for better particle management.

- **Bug Fixes**
  - Improved error handling and validation processes in various components to ensure robustness.

- **Documentation**
  - Updated documentation to reflect changes in parameter names and functionality across several modules.

- **Chores**
  - Removed unused include statements and optimized dependencies in various files to streamline the codebase.

- **Tests**
  - Updated tests to validate new output naming conventions and ensure compatibility with the revised particle handling logic.

### CODE REVIEW

File: Examples/Python/python/acts/examples/simulation.py [hunk 1]
```diff
@@ -493,45 +483,38 @@ def addFatras(
...
-        s.addWhiteboardAlias("particles_selected", particlesInitial)
-    else:
-        particlesInitial = outputParticlesInitial
-        particlesFinal = outputParticlesFinal
-
-    # Only add alias for 'particles_initial' as this is the one we use most
-    s.addWhiteboardAlias("particles", outputParticlesInitial)
-
     # Output
     addSimWriters(
```
Comment (benjaminhuth) [inline Examples/Python/python/acts/examples/simulation.py:487]: Hmm this changes which particles are written, right? before only teh selection was written, now we write before the selection. Is there a reason why this is changed?

File: Examples/Python/python/acts/examples/simulation.py [hunk 2]
```diff
@@ -738,36 +710,30 @@ def addGeant4(
     # Sequencer
     s.addAlgorithm(alg)
 
+    # Output
+    addSimWriters(
```
Comment (benjaminhuth) [inline Examples/Python/python/acts/examples/simulation.py:714]: same question as above

**Review by benjaminhuth — COMMENTED** [review]
Good change, makes some things less ambiguous and tidious! Just a few comments for now on minor things

File: Examples/Python/python/acts/examples/simulation.py (hunk 1 above)
Comment (andiwand) [reply Examples/Python/python/acts/examples/simulation.py:487]: Good point! I will revert this here as it does not need to be coupled with this change.

I was in doubt if we should write the selected or raw particles. At the same time we write all the hits so I feel like we are loosing information?

I believe the right solution is to elevate the particle selection into the user's chain and not have it in the simulation.

File: Examples/Algorithms/Fatras/src/FatrasSimulation.cpp [hunk 3]
```diff
@@ -293,9 +295,24 @@
...
+    if (auto it = particlesFinal.find(particleInitial.particleId());
+        it != particlesFinal.end()) {
+      particleSimulated.final() = *it;
+    } else {
+      ACTS_ERROR("particle " << particleInitial.particleId()
+                             << " has no final state");
+    }
+
+    particlesSimulated.insert(particleSimulated);
+  }
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Algorithms/Fatras/src/FatrasSimulation.cpp:312]: final state will be set by `particleInitial` for now

File: Examples/Algorithms/Fatras/src/FatrasSimulation.cpp [hunk 4]
```diff
@@ -263,19 +261,23 @@
...
+                                           particlesFinalUnordered.end());
   SimHitContainer simHits(simHitsUnordered.begin(), simHitsUnordered.end());
 #else
   // working around a nasty boost bug
   // https://github.com/boostorg/container/issues/244
 
-  SimParticleContainer particlesInitial;
-  SimParticleContainer particlesFinal;
+  SimParticleStateContainer particlesInitial;
+  SimParticleStateContainer particlesFinal;
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Algorithms/Fatras/src/FatrasSimulation.cpp:280]: Don't want to touch this for now as it is only a fallback. We can drop this as soon as we have a higher boost version as a requirement.

File: Examples/Algorithms/Fatras/src/FatrasSimulation.cpp [hunk 5]
```diff
@@ -263,19 +261,23 @@ ActsExamples::ProcessCode ActsExamples::FatrasSimulation::execute(
              << " simulated particles (final state)");
   ACTS_DEBUG(simHitsUnordered.size() << " simulated hits");
 
+  if (particlesInitialUnordered.size() != particlesFinalUnordered.size()) {
+    ACTS_ERROR("number of initial and final state particles differ");
+  }
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Algorithms/Fatras/src/FatrasSimulation.cpp:266]: `ACTS_ERROR` might trigger an abort depending on the user's choice.

File: Examples/Io/EDM4hep/src/EDM4hepUtil.cpp [hunk 6]
```diff
@@ -28,34 +24,33 @@ using namespace Acts::UnitLiterals;
...
+  to.initial().setDirection(momentum.normalized());
 
-  to.setAbsoluteMomentum(momentum.norm() * 1_GeV);
+  to.initial().setAbsoluteMomentum(momentum.norm() * 1_GeV);
 
   return to;
 }
 
-void EDM4hepUtil::writeParticle(const ActsFatras::Particle& from,
+void EDM4hepUtil::writeParticle(const SimParticle& from,
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Io/EDM4hep/src/EDM4hepUtil.cpp:53]: Not relevant for now

File: Examples/Io/Root/src/RootParticleReader.cpp [hunk 7]
```diff
@@ -136,18 +147,30 @@
...
+                              (*m_vt).at(i) * Acts::UnitConstants::mm);
     // NOTE: direction is normalized inside `setDirection`
-    p.setDirection((*m_px)[i], (*m_py)[i], (*m_pz)[i]);
-    p.setAbsoluteMomentum((*m_p)[i] * Acts::UnitConstants::GeV);
+    initialState.setDirection((*m_px).at(i), (*m_py).at(i), (*m_pz).at(i));
+    initialState.setAbsoluteMomentum((*m_p).at(i) * Acts::UnitConstants::GeV);
+
+    SimParticleState& finalState = p.final();
+
+    // TODO eloss cannot be read since we need the final momentum
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Io/Root/src/RootParticleReader.cpp:168]: Story for another day

File: Examples/Io/Root/src/RootParticleReader.cpp [hunk 8]
```diff
@@ -109,6 +114,12 @@ RootParticleReader::~RootParticleReader() {
   delete m_particle;
   delete m_generation;
   delete m_subParticle;
+
+  delete m_eLoss;
+  delete m_pathInX0;
+  delete m_pathInL0;
+  delete m_numberOfHits;
+  delete m_outcome;
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Examples/Io/Root/src/RootParticleReader.cpp:122]: Root things

**Review by benjaminhuth — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [69031ba9caf5184341e9e320a48b097af1f255a5]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811783)||
|:green_circle: | [run_art_test: test_run4_acts_vertex_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811782)||
|:green_circle: | [run_art_test: test_run4_ttbar_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811781)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811780)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811779)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811778)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811777)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811776)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811775)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811774)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811773)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811772)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811771)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811770)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811769)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811768)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811767)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811766)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811765)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811764)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811763)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811762)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811761)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811760)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811759)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811758)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811757)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811756)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811755)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811754)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811753)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811752)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811751)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811750)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45811749)||

**Pull request merged**